File: src/contracts/workspace/accounts.ts
import { createHash } from 'node:crypto';
import { fromJSON, get, has, int, object, set, string, same, JobsError } from './values.js';
import type { Document, Value } from './values.js';
import { exact, optionalEmail, revisionAndTime } from './automation.js';
import { emailFlow, passwordFlow } from './account-realm.js';

const legacy = ['realmRef', 'adapterId', 'descriptorVersion', 'descriptor', 'signupEmailOverride', 'providerId', 'credentialRef', 'credentialVersion', 'lifecycleState', 'revision', 'createdAt', 'updatedAt'];
const lifecycles = ['discovered', 'credential_provisioned', 'signup_in_progress', 'active', 'verification_required', 'reset_required', 'failed_definitive', 'ambiguous'];
export function validateAccount(key: string, value: Value): Document {
  const record = object(value, 'employer account');
  exact(record, record.size === legacy.length ? legacy : [...legacy, 'flowKind', 'credentialRequired'], 'employer account record is invalid');
  if (string(get(record, 'realmRef')) !== key) throw new JobsError('employer account record is invalid');
  if (!/^[0-9a-f]{64}$/.test(key)) throw new JobsError('employer account realm reference is invalid');
  const adapter = string(get(record, 'adapterId')), descriptor = string(get(record, 'descriptor'));
  const flow = has(record, 'flowKind') ? string(get(record, 'flowKind')) : passwordFlow;
  const required = has(record, 'credentialRequired') ? get(record, 'credentialRequired') : true;
  if (!['workday', 'oracle-recruiting'].includes(adapter ?? '') || flow !== (adapter === 'workday' ? passwordFlow : emailFlow)
    || required !== (adapter === 'workday') || !same(get(record, 'descriptorVersion'), fromJSON(1))
    || !descriptor?.startsWith(`${adapter}:v1:`) || createHash('sha256').update(descriptor).digest('hex') !== key) throw new JobsError('employer account realm descriptor is invalid');
  optionalEmail(get(record, 'signupEmailOverride'), 'signup email override');
  const provider = get(record, 'providerId'), reference = get(record, 'credentialRef'), version = get(record, 'credentialVersion');
  const lifecycle = string(get(record, 'lifecycleState')) ?? '';
  if (!lifecycles.includes(lifecycle)) throw new JobsError('account lifecycle state is invalid');
  if (provider === null) {
    const allowed = flow === emailFlow ? ['discovered', 'signup_in_progress', 'active', 'verification_required', 'failed_definitive', 'ambiguous'] : ['discovered', 'signup_in_progress', 'ambiguous'];
    if (reference !== null || version !== null || !allowed.includes(lifecycle)) throw new JobsError('credential metadata requires the protected provider');
  } else if (flow === emailFlow) throw new JobsError('email-only account cannot have protected credential metadata');
  else if (!/^[a-z][a-z0-9-]{2,63}$/.test(string(provider) ?? '') || !/^credential_[0-9a-f]{64}$/.test(string(reference) ?? '') || (int(version) ?? 0n) < 1n || lifecycle === 'discovered') throw new JobsError('protected credential metadata is invalid');
  revisionAndTime(record, 'employer account');
  return record;
}
export function validateAccountsDocument(value: Value): Document {
  const document = object(value, 'employer accounts');
  if (int(get(document, 'schemaVersion')) !== 1n) throw new JobsError('employer accounts schema version is unsupported');
  exact(document, ['schemaVersion', 'accounts', 'metadata'], 'employer accounts document contains unsupported fields');
  const accounts = object(get(document, 'accounts'), 'employer accounts');
  const metadata = object(get(document, 'metadata'), 'employer account metadata');
  exact(metadata, ['createdAt', 'updatedAt'], 'employer account metadata is invalid');
  for (const field of ['createdAt', 'updatedAt']) if (!string(get(metadata, field))) throw new JobsError('employer account metadata timestamp is invalid');
  for (const [key, record] of accounts.entries()) validateAccount(string(key)!, record);
  return document;
}
export function publicAccount(record: Document): Document {
  const result = object(fromJSON({}), 'account projection');
  for (const field of ['realmRef', 'adapterId', 'descriptorVersion', 'lifecycleState', 'credentialVersion', 'revision', 'createdAt', 'updatedAt']) set(result, field, get(record, field));
  set(result, 'flowKind', has(record, 'flowKind') ? get(record, 'flowKind') : fromJSON(passwordFlow));
  set(result, 'credentialRequired', has(record, 'credentialRequired') ? get(record, 'credentialRequired') : true);
  set(result, 'signupEmailOverrideConfigured', get(record, 'signupEmailOverride') !== null);
  return set(result, 'providerAssigned', get(record, 'providerId') !== null);
}

File: scripts/job_apply_store/validation/accounts.py
"""Pure automation-settings and employer-account record validation."""

from __future__ import annotations

import hashlib
import re
from typing import Any

from ..constants import EMAIL_PATTERN
from ..errors import StoreError
from ..io import require_object


def _optional_email(value: Any, label: str) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str):
        raise StoreError(f"{label} must be an email address or null")
    normalized = value.strip()
    if len(normalized) > 254 or EMAIL_PATTERN.fullmatch(normalized) is None:
        raise StoreError(f"{label} must be a valid email address")
    return normalized


def _validate_automation_settings_record(
    value: Any, *, accounts_module: Any,
) -> dict[str, Any]:
    record = require_object(value, "automation settings")
    expected = {
        "enabled", "automaticAccountCreation", "signupEmail", "passwordStrategy",
        "revision", "createdAt", "updatedAt",
    }
    if set(record) != expected:
        raise StoreError("automation settings contain unsupported fields")
    if not isinstance(record["enabled"], bool) or not isinstance(
        record["automaticAccountCreation"], bool
    ):
        raise StoreError("automation settings switches must be booleans")
    _optional_email(record["signupEmail"], "signup email")
    if record["passwordStrategy"] not in accounts_module.PASSWORD_STRATEGIES:
        raise StoreError("password strategy is unsupported")
    revision = record["revision"]
    if not isinstance(revision, int) or isinstance(revision, bool) or revision < 1:
        raise StoreError("automation settings revision must be a positive integer")
    for field in ("createdAt", "updatedAt"):
        if not isinstance(record[field], str) or not record[field]:
            raise StoreError("automation settings timestamp is invalid")
    return record


def _validate_employer_account_record(
    key: str,
    value: Any,
    *,
    accounts_module: Any,
    credentials_module: Any,
) -> dict[str, Any]:
    record = require_object(value, "employer account")
    legacy_expected = {
        "realmRef", "adapterId", "descriptorVersion", "descriptor",
        "signupEmailOverride", "providerId", "credentialRef", "credentialVersion",
        "lifecycleState", "revision", "createdAt", "updatedAt",
    }
    expected = legacy_expected | {"flowKind", "credentialRequired"}
    if set(record) not in (legacy_expected, expected) or record.get("realmRef") != key:
        raise StoreError("employer account record is invalid")
    if not re.fullmatch(r"[0-9a-f]{64}", key):
        raise StoreError("employer account realm reference is invalid")
    descriptor = record["descriptor"]
    adapter_id = record["adapterId"]
    flow_kind = record.get("flowKind", accounts_module.FLOW_PASSWORD)
    credential_required = record.get("credentialRequired", True)
    descriptor_prefix = f"{adapter_id}:v{accounts_module.REALM_DESCRIPTOR_VERSION}:"
    if (
        adapter_id not in {"workday", "oracle-recruiting"}
        or flow_kind
        != {
            "workday": accounts_module.FLOW_PASSWORD,
            "oracle-recruiting": accounts_module.FLOW_EMAIL_ONLY,
        }[adapter_id]
        or credential_required is not (adapter_id == "workday")
        or record["descriptorVersion"] != accounts_module.REALM_DESCRIPTOR_VERSION
        or not isinstance(descriptor, str)
        or not descriptor.startswith(descriptor_prefix)
        or hashlib.sha256(descriptor.encode("utf-8")).hexdigest() != key
    ):
        raise StoreError("employer account realm descriptor is invalid")
    _optional_email(record["signupEmailOverride"], "signup email override")
    provider_id = record["providerId"]
    credential_ref = record["credentialRef"]
    credential_version = record["credentialVersion"]
    lifecycle = record["lifecycleState"]
    if lifecycle not in accounts_module.LIFECYCLE_STATES:
        raise StoreError("account lifecycle state is invalid")
    if provider_id is None:
        provider_free_states = (
            {
                "discovered",
                "signup_in_progress",
                "active",
                "verification_required",
                "failed_definitive",
                "ambiguous",
            }
            if flow_kind == accounts_module.FLOW_EMAIL_ONLY
            else {"discovered", "signup_in_progress", "ambiguous"}
        )
        if (
            credential_ref is not None
            or credential_version is not None
            or lifecycle not in provider_free_states
        ):
            raise StoreError("credential metadata requires the protected provider")
    elif flow_kind == accounts_module.FLOW_EMAIL_ONLY:
        raise StoreError("email-only account cannot have protected credential metadata")
    elif (
        not isinstance(provider_id, str)
        or re.fullmatch(r"[a-z][a-z0-9-]{2,63}", provider_id) is None
        or not isinstance(credential_ref, str)
        or credentials_module.OPAQUE_REF.fullmatch(credential_ref) is None
        or not isinstance(credential_version, int)
        or isinstance(credential_version, bool)
        or credential_version < 1
        or lifecycle == "discovered"
    ):
        raise StoreError("protected credential metadata is invalid")
    revision = record["revision"]
    if not isinstance(revision, int) or isinstance(revision, bool) or revision < 1:
        raise StoreError("employer account revision must be a positive integer")
    for field in ("createdAt", "updatedAt"):
        if not isinstance(record[field], str) or not record[field]:
            raise StoreError("employer account timestamp is invalid")
    return record
